kotlinx-coroutines-core/common/src/sync/Semaphore.kt

```diff
@@ -0,0 +1,183 @@
+    override val availablePermits: Int get() = max(_availablePermits.value, 0)
+
+    // The queue of waiting acquirers is essentially an infinite array based on `SegmentQueue`;
+    // each segment contains a fixed number of slots. To determine a slot for each enqueue
+    // and dequeue operation, we increment the corresponding counter at the beginning of the operation
+    // and use the value before the increment as a slot number. This way, each enqueue-dequeue pair
+    // works with an individual cell.
+    private val enqIdx = atomic(0L)
+    private val deqIdx = atomic(0L)
+
+    override fun tryAcquire(): Boolean {
+        _availablePermits.loop { p ->
+            if (p <= 0) return false
+            if (_availablePermits.compareAndSet(p, p - 1)) return true
+        }
+    }
+
+    override suspend fun acquire() {
+        val p = _availablePermits.getAndDecrement()
+        if (p > 0) return // permit acquired
+        addToQueueAndSuspend()
+    }
+
+    override fun release() {
+        val p = _availablePermits.getAndUpdate { cur ->
+            check(cur < maxPermits) { "The number of acquired permits cannot be greater than maxPermits" }
+            cur + 1
+        }
+        if (p >= 0) return // no waiters
+        resumeNextFromQueue()
+    }
+
+    private suspend fun addToQueueAndSuspend() = suspendAtomicCancellableCoroutine<Unit> sc@ { cont ->
+        val last = this.last
+        val enqIdx = enqIdx.getAndIncrement()
+        val segment = getSegment(last, enqIdx / SEGMENT_SIZE)
+        val i = (enqIdx % SEGMENT_SIZE).toInt()
+        if (segment === null || segment[i].value === RESUMED || !segment[i].compareAndSet(null, cont)) {
+            // already resumed
+            cont.resume(Unit)
+            return@sc
+        }
+        cont.invokeOnCancellation(handler = object : CancelHandler() {
+            override fun invoke(cause: Throwable?) {
+                segment.cancel(i)
+                release()
+            }
+        }.asHandler)
+    }
+
+    private fun resumeNextFromQueue() {
+        val first = this.first
+        val deqIdx = deqIdx.getAndIncrement()
+        val segment = getSegmentAndMoveFirst(first, deqIdx / SEGMENT_SIZE) ?: return
+        val i = (deqIdx % SEGMENT_SIZE).toInt()
+        val cont = segment[i].getAndUpdate {
+            if (it === CANCELLED) CANCELLED else RESUMED
+        }
+        if (cont === null) return // just resumed
+        if (cont === CANCELLED) return // Cancelled continuation invokes `release`
+                                       // and resumes next suspended acquirer if needed.
+        cont as CancellableContinuation<Unit>
+        cont.resume(Unit)
+    }
+}
+
+private class SemaphoreSegment(id: Long, prev: SemaphoreSegment?): Segment<SemaphoreSegment>(id, prev) {
+    private val acquirers = atomicArrayOfNulls<Any?>(SEGMENT_SIZE)
+
+    operator fun get(index: Int): AtomicRef<Any?> = acquirers[index]
```

Comment:
It is not supported by atomicfu. You cannot have function returning atomicrefs. Make it return value instead and encapsulate CAS with index in a separate method. Rebase to the latest `develop` branch to see those errors. You'll also need to update `atomicfu_version=0.12.6` in `gradle.properties` as it has fixes for arrays support.